Elasticsearch exporter: event publishing

urso opened the pull request:
**Description:** Implement per document retry behavior and a set of tests verifying that the retry behavior is correctly configured and executed in conjunction with the BulkIndexer.

The change introduces helpers for creating a fake Elasticsearch endpoint that accepts bulk requests. This allows us to test the retry behavior for different failure scenarios.

**Link to tracking Issue:** #1800 

**Testing:** `TestExporter_PushEvent` tests the retry behavior for multiple error scenarios like http requests failing, or single documents being rejected due to overload.

**Documentation:** None. No user visible changes

Commit:
Elasticsearch exporter: event publishing

Implement per document retry behavior and a set of tests verifying that
the retry behavior is correctly configured and executed in conjunction
with the BulkIndexer.

The change introduces helpers for creating a fake Elasticsearch endpoint
that accepts bulk requests. This allows us to test the retry behavior
in case of multiple scenarios.

Commit:
add missing header

Commit:
rename documentAction to itemRequest

Commit:
typo

gramidt commented:
Codecov should pass once #2391 and #2389 are merged.

urso commented:
hm, CI failed in unrelated unit test:

```
ok  	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awsxrayreceiver/internal/translator	0.079s	coverage: 97.7% of statements
--- FAIL: TestSocketReadIrrecoverableNetError (0.01s)
    obsreporttest.go:160: 
        	Error Trace:	obsreporttest.go:160
        	            				obsreporttest.go:113
        	            				poller_test.go:301
        	Error:      	could not find tags
        	Test:       	TestSocketReadIrrecoverableNetError
        	Messages:   	wantTags: [{{receiver} TestSocketReadIrrecoverableNetError} {{transport} udp}] in rows []
FAIL
```

CircleCI does not allow me to rerun the test (without pushing a change)...

gramidt commented:
@urso - Yeah, not being able to re-run the test from CircleCI is a pain point; however, you can `--amend --allow-empty` to your previous commit and force push to re-kick it off.

Commit:
rerun CI

tigrannajaryan commented:
@axw @simitt @jalvz you are codeowners of elasticexporter. Please review this PR.

@urso please add @axw @simitt @jalvz (or other Elastic contributos) as codeowners for elasticsearchexporter directory.

urso commented:
> (or other Elastic contributos) as codeowners for elasticsearchexporter directory.

PR adding Codeowners for the elasticsearchexporter directory: #2462

axw commented:
I will leave it to the exporter/elasticsearchexporter codeowners to review. @urso please let me know if you'd like additional review.

faec commented on exporter/elasticsearchexporter/exporter.go:
```diff
@@ -132,9 +200,9 @@ func newElasticsearchClient(logger *zap.Logger, config *Config) (*esClientCurren
-		MaxRetries:           config.Retry.MaxRequests,
+		DisableRetry:         !config.Retry.Enabled,
+		EnableRetryOnTimeout: !config.Retry.Enabled,
+		MaxRetries:           maxAttempts,
```
Naming clarity: I generally think of "attempts" and "retries" as being off by 1 -- i.e. if the max "attempts" is 3, that means you only "retry" (try again) twice. This convention isn't universal, but it seems confusing to use both terms in close proximity. It would be nice to stick with the same term in both structs and local variables if possible, or to note the discrepancy at the definition sites if not.

urso replied:
The structs is not defined in the exporter but the go-elasticsearch dependency. Debugging the tests I learned that the library starts the count with 1 (if 0 is passed the library will default to 3 retries instead of disabling retries). Will add a clarifying comment to `maxAttempts`. After testing I figured that the behavior is more similar to an attempt, which is what I named the variable with.

faec commented on exporter/elasticsearchexporter/exporter.go:
```diff
@@ -72,17 +90,62 @@ func (e *elasticsearchExporter) pushLogsData(ctx context.Context, ld pdata.Logs)
+	// selective ACKing in the bulk response.
+	item.OnFailure = func(ctx context.Context, item esBulkIndexerItem, resp esBulkIndexerResponseItem, err error) {
+		switch {
+		case e.maxRetries < 0 || attempts < e.maxRetries && shouldRetryEvent(resp.Status):
```
It's not obvious to me how `maxRetries` and `Retry.Enabled` interact. Is it possible to have `Retry.Enabled` set to false while still setting `maxRetries > 1`? If so, there should probably be a check for that here, or a check in `newElasticSearchClient` that overwrites the configured `maxRetries` if `Retry.Enabled` is false. If not, a comment explaining why this can't arise would be nice :-)

urso replied:
if `Retry.Enabled` is false, then events should be dropped immediately on error. I used to have a check somewhere. Let me check and maybe add more tests if necessary to verify `Retry.Enabled=false` really disables all retry attempts.

urso replied:
renaming `maxRetries` to `maxAttempts`. The `maxAttempts` is actually configured in `newExporter`. In case retry is disabled we set `maxAttempts = 1`. On Failure the predicate always evaluates to false thanks to the proper initialization of `maxAttempts`.

Note: the newElasticsearchClient configures the HTTP client. If the http request did succeed, we will have the bulk indexer report the per item status to `OnSuccess` and `OnFailure`. The client handles the HTTP retries for us, but the per item retries we have to handle ourselves.

faec reviewed (commented):
Looks good to me, some small comments about naming and the details of the retry logic

urso commented on exporter/elasticsearchexporter/exporter.go:
```diff
@@ -132,9 +200,9 @@ func newElasticsearchClient(logger *zap.Logger, config *Config) (*esClientCurren
-		EnableRetryOnTimeout: true,
-		MaxRetries:           config.Retry.MaxRequests,
+		DisableRetry:         !config.Retry.Enabled,
+		EnableRetryOnTimeout: !config.Retry.Enabled,
```
Nice catch. Yes, the second should not be not.

Commit:
review + more retry disabled tests

Commit:
test

retrigger CI after unrelated test failure on windows:

```
--- FAIL: Test_statsdreceiver_EndToEnd (0.00s)
    --- FAIL: Test_statsdreceiver_EndToEnd/default_config_with_9s_interval (0.00s)
        receiver_test.go:136:
            	Error Trace:	receiver_test.go:136
            	Error:      	Received unexpected error:
            	            	listen udp 127.0.0.1:50633: bind: An attempt was made to access a socket in a way forbidden by its access permissions.
            	Test:       	Test_statsdreceiver_EndToEnd/default_config_with_9s_interval
FAIL
FAIL	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/statsdreceiver
```
